### PR TITLE
Layout object simplification

### PR DESCRIPTION
**Pull request by 1ucian0**
The layout object is going to be used as a way to define which part of the coupling map is going to be used for the mappers in general, and the swappers in particular.

This means, if you are using physical qubit 2, they assumption should not be that physical qubits 0 and 1 are available. This PR fixes that and, by that, simplifies the object structure.

### CODE REVIEW

File: qiskit/mapper/_layout.py [hunk 1]
```diff
@@ -48,62 +50,70 @@ def from_list(self, input_list):
...
+        else:
+            raise LayoutError('The map (%s -> %s) has to be a ((Register, integer) -> integer)'
+                              ' or the other way around.' % (key, value))
+
+        self._set_type_checked_item(virtual, physical)
+
+    def _set_type_checked_item(self, virtual, physical):
+        if physical in self._p2v:
+            del self._p2v[physical]
+        if virtual in self._p2v:
```
Comment (kdk) [inline qiskit/mapper/_layout.py]: `if virtual in self._v2p:`?

Comment (kdk) [inline qiskit/mapper/_layout.py]: `if virtual in self._p2v:`?

File: qiskit/mapper/_layout.py [hunk 2]
```diff
@@ -48,62 +50,70 @@ def from_list(self, input_list):
...
+            del self._v2p[virtual]
+
```
Comment (kdk) [inline qiskit/mapper/_layout.py]: Is there a risk of leaving a dangling pointer here? e.g. if `layout = Layout({ (q,0): 0 })` and I `layout[(q,0)] = 1`. What will `layout[0]` return?

File: qiskit/mapper/_layout.py [hunk 3]
```diff
@@ -48,62 +50,70 @@ def from_list(self, input_list):
             [(QuantumRegister(3, 'qr'), 0), None,
              (QuantumRegister(3, 'qr'), 2), (QuantumRegister(3, 'qr'), 3)]
         """
-        for key, value in enumerate(input_list):
-            self[key] = value
+        for physical, virtual in enumerate(input_list):
+            if Layout.is_virtual(virtual):
```
Comment (kdk) [inline qiskit/mapper/_layout.py]: Maybe raise an exception if virtual is invalid, or is there a case where we would want to drop it?

File: qiskit/mapper/_layout.py (hunk 1 above)
Comment (1ucian0) [reply qiskit/mapper/_layout.py]: good catch.

File: qiskit/mapper/_layout.py (hunk 2 above)
Comment (1ucian0) [reply qiskit/mapper/_layout.py]: Indeed. I added a test to check that out. But I have the feeling that my solution can be more efficient

File: qiskit/mapper/_layout.py (hunk 3 above)
Comment (1ucian0) [reply qiskit/mapper/_layout.py]: good one. Done.

**Pull request merged**